grpc-testtool/benches/insert.rs

```diff
@@ -0,0 +1,240 @@
+    // same runtime
+    let runtime = tokio::runtime::Runtime::new().expect("tokio startup");
+
+    // clean up anything that was running before, and make sure we have an empty directory
+    // to run the tests in
+    reset_everything().expect("unable to reset everything");
+
+    // We want a consistent seed, but we need different values for each batch, so we
+    // reseed each time we compute more data from the next seed value upwards
+    let seed = RefCell::new(0);
+
+    criterion.bench_with_input(
+        BenchmarkId::new("insert", BATCHSIZE),
+        &BATCHSIZE,
+        |b, &s| {
+            b.to_async(&runtime).iter_batched(
+                || async {
+                    let client: &Result<
+                        DatabaseClient<tonic::transport::channel::Channel>,
+                        tonic::transport::Error,
+                    > = client.get_or_init(DatabaseClient::connect(TESTURI)).await;
+                    let client = client.as_ref().expect("connection succeeded");
+
+                    // seed a new random number generator to generate data
+                    // each time we call this, we increase the seed by 1
+                    // this gives us different but consistently different random data
+                    let seed = {
+                        let mut inner = seed.borrow_mut();
+                        *inner += 1;
+                        *inner
+                    };
+                    let mut rng = rand::rngs::StdRng::seed_from_u64(seed);
+
+                    // generate the put request, which is BATCHSIZE PutRequest objects with random keys/values
+                    let put_requests: Vec<PutRequest> = (0..s)
+                        .map(|_| {
+                            (
+                                rng.borrow_mut()
+                                    .sample_iter(&Alphanumeric)
+                                    .take(KEYLEN)
+                                    .collect::<Vec<u8>>(),
+                                rng.borrow_mut()
+                                    .sample_iter(&Alphanumeric)
+                                    .take(DATALEN)
+                                    .collect::<Vec<u8>>(),
+                            )
+                        })
+                        .map(|(key, value)| PutRequest { key, value })
+                        .collect();
+
+                    // wrap it into a tonic::Request to pass to the execution routine
+                    let req = tonic::Request::new(WriteBatchRequest {
+                        puts: put_requests,
+                        deletes: vec![],
+                    });
+
+                    // hand back the client and the request contents to the benchmark executor
+                    (client.clone(), req)
+                },
+                // this part is actually timed
+                |wbr| async {
+                    let (mut client, req) = wbr.await;
```

Comment:
Does calling `await` here mean that we are including the time used to generate the inputs in the benchmark? When is the body of the async `setup` evaluated?

Reply:
Yes! Good catch.

Fixed in latest version.